erts/emulator/nifs/win32/win_prim_file.c

```diff
@@ -716,39 +716,90 @@ static int is_executable_file(const efile_path_t *path) {
-                                  &unused_length, NULL);
-
-        /* ERROR_MORE_DATA is tolerated since we're guaranteed to have filled
-         * the field we want. */
-        if(success || GetLastError() == ERROR_MORE_DATA) {
-            result = IsReparseTagNameSurrogate(reparse_buffer.ReparseTag);
+    /* ERROR_MORE_DATA is tolerated since we're guaranteed to have filled
+     * the field we want. */
+    if(success || GetLastError() == ERROR_MORE_DATA) {
+        return IsReparseTagNameSurrogate(reparse_buffer.ReparseTag);
+    }
+
+    return 0;
+}
+
+static void build_file_info(BY_HANDLE_FILE_INFORMATION *native_file_info, const efile_path_t *path, int is_link, efile_fileinfo_t *result) {
+    DWORD attributes;
+
+    attributes = native_file_info->dwFileAttributes;
+
+    if(is_link) {
+        result->type = EFILE_FILETYPE_SYMLINK;
+        /* This should be _S_IFLNK, but the old driver always set
+         * non-directories to _S_IFREG. */
+        result->mode |= _S_IFREG;
+    } else if(attributes & FILE_ATTRIBUTE_DIRECTORY) {
+        result->type = EFILE_FILETYPE_DIRECTORY;
+        result->mode |= _S_IFDIR | _S_IEXEC;
+    } else {
+        if(is_executable_file(path)) {
+            result->mode |= _S_IEXEC;
         }
 
-        CloseHandle(handle);
-     }
+        result->type = EFILE_FILETYPE_REGULAR;
+        result->mode |= _S_IFREG;
+    }
 
-     return result;
+    if(!(attributes & FILE_ATTRIBUTE_READONLY)) {
+        result->access = EFILE_ACCESS_READ | EFILE_ACCESS_WRITE;
+        result->mode |= _S_IREAD | _S_IWRITE;
+    } else {
+        result->access = EFILE_ACCESS_READ;
+        result->mode |= _S_IREAD;
+    }
+
+    /* Propagate user mode-bits to group/other fields */
+    result->mode |= (result->mode & 0700) >> 3;
+    result->mode |= (result->mode & 0700) >> 6;
+
+    result->size =
+        ((Uint64)native_file_info->nFileSizeHigh << 32ull) |
+        (Uint64)native_file_info->nFileSizeLow;
+    result->links = MAX(1, native_file_info->nNumberOfLinks);
+
+    result->major_device = get_drive_number(path);
+    result->minor_device = 0;
+    result->inode = 0;
+    result->uid = 0;
+    result->gid = 0;
+}
+
+static void build_file_info_times(BY_HANDLE_FILE_INFORMATION *native_file_info, efile_fileinfo_t *result) {
+    FILETIME_TO_EPOCH(result->m_time, native_file_info->ftLastWriteTime);
+    FILETIME_TO_EPOCH(result->a_time, native_file_info->ftLastAccessTime);
+    FILETIME_TO_EPOCH(result->c_time, native_file_info->ftCreationTime);
+
+    if(result->m_time == -EPOCH_DIFFERENCE) {
+        /* Default to 1970 just like the old driver. */
+        result->m_time = 0;
+    }
+
+    if(result->a_time == -EPOCH_DIFFERENCE) {
+        result->a_time = result->m_time;
+    }
+
+    if(result->c_time == -EPOCH_DIFFERENCE) {
+        result->c_time = result->m_time;
+    }
 }
 
 posix_errno_t efile_read_info(const efile_path_t *path, int follow_links, efile_fileinfo_t *result) {
```

Comment:
`attributes` need to be used for the root path fallbacks to work.

```diff
-        attributes = FILE_ATTRIBUTE_DIRECTORY;
+        native_file_info.dwFileAttributes = FILE_ATTRIBUTE_DIRECTORY;
     } else if(is_path_root(path)) {
         /* Local (or mounted) roots can be queried with GetFileAttributesW but
          * lack support for GetFileInformationByHandle, so we'll skip that
          * part. */
+        native_file_info.dwFileAttributes = attributes;
```

We may need to do something similar for the variant that takes a handle, falling back to looking at the path if the call to `GetFileInformationByHandle` fails.

Reply:
I've fixed the `attributes` usage. About the second part, do you mean something like this in `efile_read_handle_info`?

```
posix_errno = internal_read_link(w->handle, &path);
if(posix_errno != 0) {
    return posix_errno;
}
if(!GetFileInformationByHandle(w->handle, &native_file_info)) {
    if(is_path_root(&path)) {
        // fallback
        return efile_read_info(&path, follow_links, result);
    }
    return windows_to_posix_errno(GetLastError());
}
```

Reply:
I'm a bit out of my depth here as path handling in windows is a mystery to me... Wouldn't doing it inline instead of just calling efile_read_info() require duplicating its code in efile_read_handle_info()?

Reply:
By the way, I was wondering if it wouldn’t be better to just return the error if `GetFileInformationByHandle` fails. 

If we fallback to getting the file info by path, wouldn’t it be possible for the path root to be replaced by something else after the error but before the fallback call, leaving the caller with an unexpected result (I have this descriptor, but the file info I requested is not related to it)?

Reply:
I went ahead and implemented the fallback in the last commit so it's done if you prefer to go with it.

Reply:
> If we fallback to getting the file info by path, wouldn’t it be possible for the path root to be replaced by something else after the error but before the fallback call, leaving the caller with an unexpected result (I have this descriptor, but the file info I requested is not related to it)?

A "root path" is something that explicitly points at a drive like `C:\`, a network share, or similar. Should `C:\` start to point at a different drive, e.g. if the user unplugs a USB drive and then puts in a new one, all our open handles on that device will be dead and we'll error out when trying to figure out which path the handle refers to.

> By the way, I was wondering if it wouldn’t be better to just return the error if `GetFileInformationByHandle` fails.
> 

That would make it impossible to query root paths, which would differ from `efile_read_info`.

> I'm a bit out of my depth here as path handling in windows is a mystery to me... Wouldn't doing it inline instead of just calling efile_read_info() require duplicating its code in efile_read_handle_info()?

Sorry about the back-and-forth on this; after sleeping on it it's best if you call `efile_read_info` on failure. I was thinking about just saying "it's a directory" if we have a root path and leaving it at that, but then again there's bound to be some nasty edge cases around here. I'm not comfortable with Windows path handling either. :(

Reply:
> Sorry about the back-and-forth on this; after sleeping on it it's best if you call efile_read_info on failure.

No problem! I've pushed an implementation of the fallback to `efile_read_info` which does it only for path roots. Do you think it should be done for any `GetFileInformationByHandle` errors?

Reply:
> No problem! I've pushed an implementation of the fallback to `efile_read_info` which does it only for path roots. Do you think it should be done for any `GetFileInformationByHandle` errors?

Thanks!

I don't think we should retry on all failures, roots are just an ugly special case we need to handle.

Reply:
We're still passing the path to `internal_read_link` in this function, causing `read_file_info` on links to fail.

`posix_errno = internal_read_link(path, &resolved_path);`

Reply:
I’m really sorry about that. I’ll fix it when I arrive at work and do a full test run instead of relying just on reviewing the patches.

Reply:
So, I wasn't able to run the automated tests on windows. I got an error when running the `Install` program saying that `-minimal` wasn't a recognized flag, so I edited the code to force the minimal mode without that flag, but then I got an error saying the `Install.ini` file could not be opened, although it's clearly there... So I'm not sure if something's broken in my Cygwin install.

I have tested manually, though, and I think things are working fine. I've tested links (via `ln -s` on Cygwin, shortcuts created on Windows Explorer, and the `mklink` command). I didn't test `\\localhost\c$` directly because this branch doesn't have the patch that allows `file:open` on directories. I did open a file using that kind of path (i.e. `\\localhost\c$\foo.txt`) and it works, though I'm not sure if that counts as the same test.

Reply:
> I’m really sorry about that. I’ll fix it when I arrive at work and do a full test run instead of relying just on reviewing the patches.

Don't sweat it, we're just glad to have your help. :o)

> So, I wasn't able to run the automated tests on windows. I got an error when running the `Install` program saying that `-minimal` wasn't a recognized flag, so I edited the code to force the minimal mode without that flag, but then I got an error saying the `Install.ini` file could not be opened, although it's clearly there... So I'm not sure if something's broken in my Cygwin install.

I usually run the tests in the source tree, maybe you'll have better luck with that;

```
%% in bash.
export TESTROOT=$ERL_TOP/release_tests;
make release_tests;
cd $TESTROOT/test_server;
$ERL_TOP/bin/werl;

%% in the erlang shell
ts:install().
ts:run(kernel, file_SUITE, [batch]).
```

> I have tested manually, though, and I think things are working fine. I've tested links (via `ln -s` on Cygwin, shortcuts created on Windows Explorer, and the `mklink` command). I didn't test `\\localhost\c$` directly because this branch doesn't have the patch that allows `file:open` on directories.

Feel free to squash and rebase to `master`.

> I did open a file using that kind of path (i.e. `\\localhost\c$\foo.txt`) and it works, though I'm not sure if that counts as the same test.

Sadly it doesn't, `\\localhost\c$\foo.txt` points at a file on a network share, but we want to query the share itself (`\\localhost\c$`) as if it were a directory, even though it's not.

Looking more closely at the tests it seems we're not checking this even though we should, can you add that path to `file_info_basic_directory` / `file_handle_info_basic_directory`?

Reply:
Awesome, that worked! The commits are rebased to master and squashed, and the test cases for `\\localhost\c$` were added.

Reply:
Thanks, the tests ran well last night, so I'll merge this early next week if nothing new pops up. :)

Reply:
... `file_name_SUITE:normal` went wrong on some of our machines, didn't notice it until now :(

It crashes on the first line of `rm_r` with `{error, enoent}`, it looks like there's a problem with `read_link_info`.

Reply:
Was that on Windows only?

Reply:
Hi

Is there anything in common among the failing machines? Windows version, 32 bit vs 64 arch, etc? I can't reproduce the error here on Cygwin on 64-bit Windows 10 :(